File: src/main/java/com/vaadin/demo/component/emailfield/EmailFieldValidation.java
package com.vaadin.demo.component.emailfield;

import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.textfield.EmailField;
import com.vaadin.flow.component.textfield.EmailField.EmailFieldI18n;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("email-field-validation")
public class EmailFieldValidation extends HorizontalLayout {

    public EmailFieldValidation() {
        setPadding(false);

        // tag::snippet[]
        EmailField field = new EmailField("Email address");
        field.setRequiredIndicatorVisible(true);
        field.setPattern(
                "^[a-zA-Z0-9_\\-+]+(?:\\.[a-zA-Z0-9_\\-+]+)*@example\\.com$");

        field.setI18n(new EmailFieldI18n()
                .setRequiredErrorMessage("Field is required")
                .setPatternErrorMessage(
                        "Enter a valid example.com email address"));
        // end::snippet[]
        field.setHelperText("Only example.com addresses allowed");

        add(field);
    }

    public static class Exporter extends DemoExporter<EmailFieldValidation> { // hidden-source-line
    } // hidden-source-line
}

File: frontend/demo/component/emailfield/email-field-validation.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/email-field';
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import type { EmailField, EmailFieldValidatedEvent } from '@vaadin/email-field';
import { applyTheme } from 'Frontend/demo/theme';

@customElement('email-field-validation')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  @state()
  private errorMessage = '';

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-email-field
        required
        pattern="^[a-zA-Z0-9_\\-+]+(?:\\.[a-zA-Z0-9_\\-+]+)*@example\\.com$"
        label="Email address"
        helper-text="Only example.com addresses allowed"
        .errorMessage="${this.errorMessage}"
        @validated="${(event: EmailFieldValidatedEvent) => {
          const field = event.target as EmailField;
          const { validity } = field.inputElement as HTMLInputElement;
          if (validity.valueMissing) {
            this.errorMessage = 'Field is required';
          } else if (validity.patternMismatch) {
            this.errorMessage = 'Enter a valid example.com email address';
          } else {
            this.errorMessage = '';
          }
        }}"
      ></vaadin-email-field>
      <!-- end::snippet[] -->
    `;
  }
}
